qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/QuteCodeActions.java

```diff
@@ -397,4 +441,128 @@ private static void doCodeActionsForUndefinedSectionTag(Template template, Diagn
 		}
 	}
+
+	private void doCodeActionsForSimilarValues(Template template, Diagnostic diagnostic, List<CodeAction> codeActions,
+			String varName) throws BadLocationException {
+		Range diagnosticRange = diagnostic.getRange();
+		int offset = template.offsetAt(diagnosticRange.getEnd());
+
+		Node nodeBefore = template.findNodeBefore(offset);
+		if (nodeBefore == null) {
+			return;
+		}
+
+		Node node = QutePositionUtility.findBestNode(offset, nodeBefore);
+
+		if (node.getKind() == NodeKind.ExpressionPart) {
+			Range rangeValue = new Range(
+					new Position(diagnosticRange.getStart().getLine(), diagnosticRange.getStart().getCharacter()),
+					new Position(diagnosticRange.getEnd().getLine(), diagnosticRange.getEnd().getCharacter()));
+
+			Collection<String> availableValues = new TreeSet<>();
+
+			switch (((Part) node).getPartKind()) {
+			case Object:
+				collectAvailableValuesForObjectPart(node, template, availableValues);
+			case Namespace:
+				collectAvailableValuesForNamespacePart(node, template, availableValues);
+				break;
+			case Method:
+			case Property:
+				String signature = null;
+				JsonObject data = (JsonObject) diagnostic.getData();
+				if (data != null) {
+					signature = data.get(DIAGNOSTIC_DATA_TAG).getAsString();
+					collectAvailableValuesForJavaTypePart(node, template, signature, availableValues);
+					break;
+				}
+			}
+
+			for (String value : availableValues) {
+				if (isSimilar(value, varName)) {
+					CodeAction similarCodeAction = CodeActionFactory.replace("Did you mean '" + value + "'?",
+							rangeValue, value, template.getTextDocument(), diagnostic);
+					codeActions.add(similarCodeAction);
+				}
+			}
+		}
+	}
+
+	private void collectAvailableValuesForObjectPart(Node node, Template template, Collection<String> availableValues) {
+		String projectUri = template.getProjectUri();
+
+		List<ValueResolver> globalResolvers = javaCache.getGlobalVariables(projectUri);
+		for (ValueResolver resolver : globalResolvers) {
+			availableValues.add(resolver.getName());
+		}
+
+		List<String> aliases = template.getChildren().stream() //
+				.filter(n -> n.getKind() == NodeKind.ParameterDeclaration) //
+				.map(n -> ((ParameterDeclaration) n).getAlias()) //
+				.filter(alias -> alias != null) //
+				.collect(Collectors.toList());
+		for (String alias : aliases) {
+			availableValues.add(alias);
+		}
+
+		ExtendedDataModelTemplate dataModel = javaCache.getDataModelTemplate(template).getNow(null);
+		if (dataModel != null) {
+			for (ExtendedDataModelParameter parameter : dataModel.getParameters()) {
+				availableValues.add(parameter.getKey());
+			}
+		}
+
+		Section section = node != null ? node.getParentSection() : null;
+		if (section != null) {
+			List<SectionMetadata> metadatas = section.getMetadata();
+			for (SectionMetadata metadata : metadatas) {
+				availableValues.add(metadata.getName());
+			}
+		}
+
+		Collection<SectionMetadata> specialKeysMetadatas = UserTagUtils.getSpecialKeys();

```

Comment:
`if (UserTagUtils.isUserTag(template)) {`

It will fix  `QuteDiagnosticsInExpressionTest.undefinedUnderscoreObject:205 expected: <3> but was: <4>`

Please write tests with user tag (and the same test without user tag) and metada. like https://github.com/redhat-developer/quarkus-ls/blob/8a52980f2a999e80a0367f301f8962dfaf4f84b7/qute.ls/com.redhat.qute.ls/src/test/java/com/redhat/qute/services/hover/QuteHoverInUserTag.java#L97